controllers/dockermachine_controller.go

```diff
@@ -115,6 +118,38 @@ func (r *DockerMachineReconciler) Reconcile(req ctrl.Request) (ctrl.Result, erro
 		return ctrl.Result{}, err
 	}
 
+
+	// If the DockerMachine doesn't have finalizer, add it.
+	if !util.Contains(dockerMachine.Finalizers, capiv1alpha2.MachineFinalizer) {
+		dockerMachine.Finalizers = append(dockerMachine.Finalizers, capiv1alpha2.MachineFinalizer)
+		if err := r.Client.Patch(ctx, dockerMachine, patch) ; err!= nil {
```

Comment:
Instead of doing the Patch inline like this, it might be better to add a defer after we create `patch` so that we ensure that we update the resource with any changes in all cases.

There will also need to be a call to `r.Client.Status().Patch()` to update any needed changes to the status (since they are not updated with the first patch call). Similar to: https://github.com/kubernetes-sigs/cluster-api/blob/master/pkg/controller/machine/machine_controller.go#L116-L136

Reply:
Let's hold off on the defer for now and make a function that patches the object like https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/blob/master/controllers/kubeadmconfig_controller.go#L326 so we don't run into the same problem where errors are lost and potentially swallowed like they are in CAPA

We will have to put this at every site we want a patch but i think it's preferable to be explicit about our patches so we can handle/log/return errors properly

Reply:
upon further investigation, we can totally use a defer function + a named return parameter.

https://golang.org/ref/spec#Defer_statements

This would satisfy my desire to return errors properly and also allow a defer function to do the cleanup work.

Reply:
Here is an example: https://play.golang.org/p/_ycV8jk6QTJ

Reply:
which of course is what capi is already doing :)